### PR TITLE
Random

### PR DESCRIPTION
**Pull request by arsalan-motamedi**
**Context:**
Adding random methods.
**Description of the Change:**
We have
Ket.random(modes),    DM.random(modes),    Unitary.random(modes),    Channel.random(modes)

**Benefits:**
Clear: an obvious use case is for writing tests.

**Possible Drawbacks:**
None.
**Related GitHub Issues:**
None.

### CODE REVIEW

File: mrmustard/lab_dev/states/base.py [hunk 1]
```diff
@@ -829,6 +829,21 @@ def __getitem__(self, modes: Union[int, Sequence[int]]) -> State:
             representation, wires, self.name
         )  # pylint: disable=protected-access
 
+    @classmethod
+    def random(cls, modes, max_r=1.0):
```
Comment (ziofil) [inline mrmustard/lab_dev/states/base.py:833]: Can we make `m` a parameter (that defaults to `len(m)`? That way one can make the DM more/less mixed by tracing a smaller/larger ancilla.

Comment (arsalan-motamedi) [reply mrmustard/lab_dev/states/base.py:833]: Sure thing!

File: mrmustard/lab_dev/states/base.py [hunk 2]
```diff
@@ -1043,3 +1058,38 @@ def __rshift__(self, other: CircuitComponent | Scalar) -> CircuitComponent | Bat
...
+        Args: "modes" are the modes where the state is defined on
+        Output is a Ket
+        """
+        # TODO: use __class_getitem_ and sample from broader probabilities
+        # TODO: random any gate
+
+        # generate a random ket repr.
+        # e.g. S = math.random_symplectic(dim) and then apply to vacuum
+        m = len(modes)
+        S = math.random_symplectic(m, max_r)  # change to a,a^dagger
```
Comment (ziofil) [inline mrmustard/lab_dev/states/base.py:1075]: what do you mean by change to a,a^dagger?

Comment (arsalan-motamedi) [reply mrmustard/lab_dev/states/base.py:1075]: The symplectic transform after applying \Omega = 1/sqrt(2) * [[1,1],[-i,i]]

File: mrmustard/lab_dev/states/base.py (hunk 1 above)
Comment (arsalan-motamedi) [reply mrmustard/lab_dev/states/base.py:833]: Should I do the same with Channels?

File: mrmustard/lab_dev/states/base.py [hunk 3]
```diff
@@ -1043,3 +1058,38 @@ def __rshift__(self, other: CircuitComponent | Scalar) -> CircuitComponent | Bat
...
+                [
+                    [math.eye(m, dtype=math.complex128), math.eye(m, dtype=math.complex128)],
+                    [
+                        -1j * math.eye(m, dtype=math.complex128),
+                        1j * math.eye(m, dtype=math.complex128),
+                    ],
+                ]
+            )
+        )
+        S = np.conjugate(math.transpose(Transformation)) @ S @ Transformation
```
Comment (ziofil) [inline mrmustard/lab_dev/states/base.py:1089]:
```suggestion
        S = math.conj(math.transpose(Transformation)) @ S @ Transformation
```

File: mrmustard/lab_dev/states/base.py [hunk 4]
```diff
@@ -1043,3 +1058,38 @@ def __rshift__(self, other: CircuitComponent | Scalar) -> CircuitComponent | Bat
...
+        S_1 = S[:m, :m]
+        S_2 = S[:m, m:]
+        A = S_2 @ math.conj(math.inv(S_1))  # use solve for inverse
+        b = [complex(0)] * m
```
Comment (ziofil) [inline mrmustard/lab_dev/states/base.py:1093]:
```suggestion
        b = math.zeros(m, dtype=A.dtype)
```

File: mrmustard/lab_dev/transformations/base.py [hunk 5]
```diff
@@ -168,31 +167,32 @@ def __rshift__(self, other: CircuitComponent) -> CircuitComponent:
...
+    #             "Symplectic matrix and number of modes don't match. "
+    #             + f"Modes imply shape {(M,M)}, "
+    #             + f"but shape is {symplectic.shape[-2:]}."
+    #         )
+    #     A, b, c = physics.bargmann.wigner_to_bargmann_U(symplectic, math.zeros(M))
+    #     return Unitary._from_attributes(
+    #         representation=Bargmann(A, b, c),
+    #         wires=Wires(set(), set(), set(modes_out), set(modes_in)),
+    #         name=name,
+    #     )
```
Comment (ziofil) [inline mrmustard/lab_dev/transformations/base.py:195]: you can delete this

File: mrmustard/lab_dev/states/base.py (hunk 2 above)
Comment (arsalan-motamedi) [reply mrmustard/lab_dev/states/base.py]: I removed the comment (I totally agree with you that it is not a clarifying comment).

File: mrmustard/lab_dev/states/base.py [hunk 6]
```diff
@@ -1043,3 +1058,38 @@ def __rshift__(self, other: CircuitComponent | Scalar) -> CircuitComponent | Bat
...
+        Output is a Ket
+        """
+        # TODO: use __class_getitem_ and sample from broader probabilities
+        # TODO: random any gate
+
+        # generate a random ket repr.
+        # e.g. S = math.random_symplectic(dim) and then apply to vacuum
+        m = len(modes)
+        S = math.random_symplectic(m, max_r)  # change to a,a^dagger
+        Transformation = (
```
[the remark this answers is not in the document]
Comment (arsalan-motamedi) [reply mrmustard/lab_dev/states/base.py]: Noted. I changed it throughout.

**Review by ziofil — APPROVED** [review]
awesome!

**Pull request merged**